### PR TITLE
Introduce NodeVersionInfoProvider

### PR DESCRIPTION
**Pull request by rita-sonarsource**
Two commits can be looked at separately, the first one is cosmetic renaming.
Prep for [#2997](https://github.com/SonarSource/sonarlint-visualstudio/issues/2997)

### CODE REVIEW

File: src/TypeScript/NodeJSLocator/NodeVersionInfoProvider.cs [hunk 1]
```diff
@@ -19,48 +19,42 @@
...
             Func<string, Version> getNodeExeVersion)
         {
             this.nodeLocationsProvider = nodeLocationsProvider;
-            this.logger = logger;
             this.fileSystem = fileSystem;
             this.getNodeExeVersion = getNodeExeVersion;
         }
 
-        public string Locate()
+        public IEnumerable<NodeVersionInfo> GetAllNodeVersions()
```
Comment (rita-sonarsource) [inline src/TypeScript/NodeJSLocator/NodeVersionInfoProvider.cs:57]: This class is now responsible for lazily locating all the node installations and calculating their versions.
Telemetry will call this.

File: src/TypeScript/EslintBridgeClient/EslintBridgeProcess.cs [hunk 2]
```diff
@@ -65,7 +65,7 @@ internal sealed class EslintBridgeProcess : IEslintBridgeProcess
         private static readonly object Lock = new object();
 
         private readonly string eslintBridgeStartupScriptPath;
-        private readonly INodeLocator nodeLocator;
+        private readonly ICompatibleNodeLocator compatibleNodeLocator;
```
Comment (rita-sonarsource) [inline src/TypeScript/EslintBridgeClient/EslintBridgeProcess.cs:68]: Cosmetics. Can be look at separately in the first commit.

File: src/TypeScript/NodeJSLocator/NodeVersionInfoProvider.cs [hunk 3]
```diff
@@ -71,23 +65,8 @@ public string Locate()
...
 
-                if (!IsCompatibleVersion(nodeVersion))
-                {
-                    logger.WriteLine(Resources.ERR_IncompatibleVersion, nodeVersion, nodeExePath);
-                    continue;
-                }
-
-                logger.WriteLine(Resources.INFO_FoundCompatibleVersion, nodeVersion, nodeExePath);
-                return nodeExePath;
+                yield return new NodeVersionInfo(nodeExePath, nodeVersion);
```
Comment (rita-sonarsource) [inline src/TypeScript/NodeJSLocator/NodeVersionInfoProvider.cs:68]: yield return so it doesn't degrade the performance of eslint-bridge initialization. We will still stop as soon as we find the first compatible.

File: src/TypeScript.UnitTests/NodeJSLocator/CompatibleNodeLocatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        public void Locate_NoDetectedVersions_Null()
+        {
+            var testSubject = CreateTestSubject();
+            var result = testSubject.Locate();
+
+            result.Should().BeNull();
+        }
+
+        [TestMethod]
+        public void Locate_ReturnsFirstCompatiblePath()
```
Comment (rita-sonarsource) [inline src/TypeScript.UnitTests/NodeJSLocator/CompatibleNodeLocatorTests.cs:55]: Tests are now simplified

File: src/TypeScript/NodeJSLocator/NodeVersionInfoProvider.cs (hunk 3 above)
Comment (duncanp-sonar) [reply src/TypeScript/NodeJSLocator/NodeVersionInfoProvider.cs:68]: It's a non-obvious perf optimisation, so it's worth commenting in the code.

**Review by duncanp-sonar — APPROVED** [review]
My former boss was fond of saying "there's no problem that can't be solved by adding a layer of abstraction, except having too many layers of abstraction."

Having `ICompatibleNodeLocator`, `INodeVersionInfoProvider` and `INodeLocationsProvider` and all of the associated tests is starting to feel over-complicated. Still, it's subjective whether the alternatives are any better, so I wouldn't change it.

**Review by ugras-ergun-sonarsource — APPROVED** [review]

**Pull request merged**